include/fc/log/dmlog_appender.hpp

```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include <fc/log/appender.hpp>
+
+namespace fc {
+
+   class dmlog_appender : public appender
+   {
+       public:
+            dmlog_appender( const variant& args );
+            dmlog_appender();
+
+            virtual ~dmlog_appender();
+            void initialize( boost::asio::io_service& io_service ) {
+                 setbuf(stdout, NULL);
```

Comment:
I still do not like this hijacking `stdout`. Even more so that it is part of `fc`. Can you not just call `fflush` in `dmlog_appender::log` ?

Reply:
So, our initial implementation of the `log` method was exactly ending with this in its initial implementation:

```
      FILE* out = stdout;
     
      // Format message to `message` ...

      fprintf( out, "%s\n", message.c_str() );
      fflush( out );
```

This worked for a while but a some point, we started seeing some error in the process consuming the log line emitted by this appender.

What happened is that an error occurred in the low-level I/O system, causing the line to not be printed completely, the line was truncated in the output. This was highly problematic since depending on where the truncation would happen, it could go unnoticed by the reading process, in the middle of a big hex string representing a smart contract code for example.

At that point, we tried various techniques with the standard buffered `stdout` variant, but it appeared that recovering the error and continue the writing with the remainder of the buffer is not really feasible, error recovery with buffered output is mostly impossible due to buffering being controlled by either libc or a lower level system call.

Hence, we switched to unbuffered `stdout` and we retry the write ourselves, which resolves the problem. This has been like that in production since the last year without ever seeing a truncating happening again.

Now, I do understand that it's problematic to play with the `stdout` buffering capabilities. An upcoming changes that would change from `stdout` to a `fifo` would resolve the case here.

What would be acceptable solution for you so we can have this merge and deal with the `fifo` things in the upcoming weeks?

If I move the initialize only on the first ever invocation of `log`, would that represent a good compromise for the time being?

Reply:
Please set it up so the hack of disable of stdout buffering is only done when `deep-mind = true`

Reply:
Perfect, I'll do that and add a comment explaining the reasoning and next steps.

Reply:
I moved it in `enable_deep_mind` in `controller.cpp`: https://github.com/EOSIO/eos/pull/8788/files#diff-147a7174fb705ef227e4e28b336acd4fR3335